luaui/Widgets/gfx_lupsgl4_orbs.lua

```diff
@@ -0,0 +1,748 @@
+	float opac_vs;
+	vec4 modelPos_vs;
+};
+
+//__ENGINEUNIFORMBUFFERDEFS__
+
+struct SUniformsBuffer {
+    uint composite; //     u8 drawFlag; u8 unused1; u16 id;
+
+    uint unused2;
+    uint unused3;
+    uint unused4;
+
+    float maxHealth;
+    float health;
+    float unused5;
+    float unused6;
+
+    vec4 drawPos;
+    vec4 speed;
+    vec4[4] userDefined; //can't use float[16] because float in arrays occupies 4 * float space
+};
+
+layout(std140, binding=1) readonly buffer UniformsBuffer {
+    SUniformsBuffer uni[];
+};
+
+
+#line 10468
+void main()
+{
+	if ((uni[instData.y].composite & 0x00000003u) < 1u ) { // not drawn
+		// Ivand's recommendation is to place vertices outside of NDC space:
+		gl_Position = vec4(2.0,2.0,2.0,1.0);
+		return;
+	}
+	
+	vec3 modelWorldPos = uni[instData.y].drawPos.xyz;// + vec3(100,100,100);
+	unitID_vs = 0.1 + float(uni[instData.y].composite >> 16 ) / 256000.0;
+	
+	float modelRot = uni[instData.y].drawPos.w;
+	mat3 rotY = rotation3dY(0);
+		
+	vec4 vertexWorldPos = vec4(1);
+	vec3 flippedPos = vec3(1,-1,1) * position.xzy;
+
+	
+	float radius = 0.99 * posrad.w;
+	
+	vertexWorldPos.xyz = rotY * ( flippedPos * (radius) + posrad.xyz + vec3(0,0,0) ) + modelWorldPos;
+	
+	gl_Position = cameraViewProj * vertexWorldPos;
+
+	
+
+	mat3 normalMatrix = mat3( cameraView[0].xyz, cameraView[1].xyz, cameraView[2].xyz);
+	
+	vec3 normal = (rotY * normals.xzy);
+	normal.y *= -1;
+	vec3 camPos = cameraViewInv[3].xyz;
+	vec3 camToWorldPos = normalize(cameraViewInv[3].xyz  - vertexWorldPos.xyz);
+	//vec3 vertex = vec3(gl_ModelViewMatrix * gl_Vertex);
+	color1_vs.rgb = camToWorldPos.xyz;
+	float angle = dot(normal,camToWorldPos); //*inversesqrt( dot(normal,normal)*dot(position,position) ); //dot(norm(n),norm(v))
+	opac_vs = pow( abs( angle ) , margin_teamID_shield_technique.x);
+	//opac_vs = 1.0;
+	//color1_vs.rgb = vec3(angle);
+	
+	vec4 color2_vs;
+	if (color1.r < 0) { // negative numbers mean teamcolor
+		vec4 teamcolor = teamColor[int(margin_teamID_shield_technique.y)];
+		//	ShieldSphereParticle.Default.colormap1 = {{(r*0.45)+0.3, (g*0.45)+0.3, (b*0.45)+0.3, 0.6}}
+		//ShieldSphereParticle.Default.colormap2 = {{r*0.5, g*0.5, b*0.5, 0.66} }
+		color1_vs = vec4(teamcolor.rgb, 0.5);
+		color1_vs.rgb = color1_vs.rgb * 0.45 + 0.3;
+		color2_vs = vec4(teamcolor.rgb, 0.66);
+		color2_vs.rgb = color2_vs.rgb * 0.5;
+	}else{ // base color
+		color1_vs = color1;
+		color2_vs = color2;
+	}
+	
+	color1_vs = mix(color1_vs, color2_vs, opac_vs);
+	
+	if (margin_teamID_shield_technique.z > 0.5){
+		float shieldPower = clamp(uni[instData.y].userDefined[0].z, 0, 1);
+		color1_vs = mix(vec4(0.8,0.4, 0,1.0),color1_vs,  shieldPower);
+	}
+	
+	float relHealth  = clamp(uni[instData.y].health/uni[instData.y].maxHealth, 0, 1);
+	//color1_vs.rgb *= 1.0 + relHealth;
+	
+	
+	modelPos_vs = vec4(position.xzy*posrad.w, 0);
+	//modelPos_vs.z = fract(modelPos_vs.z * 10);
+	modelPos_vs.w = relHealth;
+	gameFrame_vs = (timeInfo.x + timeInfo.w) ;
```

Comment:
```suggestion
	gameFrame_vs = timeInfo.z;
```
and then don't have to multiply by 0.0333 below (or maybe could use `timeInfo.z` directly instead of going through `gameFrame_vz`? idk)

Reply:
w is frametimeoffset, thats needed for smoothness

Reply:
i know what `w` is but apparently you don't know what `z` is

Reply:
vec4 timeInfo; //gameFrame, gameSeconds, drawFrame, frameTimeOffset
isnt it drawframe? did it get updated?

Reply:
ah I see that is has been updtaed

Reply:
yes, `z` implements the `(x+w) / 30` pattern that seems to appear all over the place